### ISSUE TITLE
Ehcache No-parameter constructor even with proper Reflection configuration

### ISSUE THREAD

**Issue by sendev1** [oracle/graal#2788]
**Describe the issue**
The project I am trying graalvm is using ehcache and when I use the native-image-agent I see that reflect-config.json is having entry for `DefaultTransactionManagerLookup ` but not sure why I am getting the exception ``net.sf.ehcache.transaction.manager.DefaultTransactionManagerLookup` can not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.` when I run the generated native image. As you can see from the refection config, the class DefaultTransactionManagerLookup has an entry and I expect this should work.

```
{
  "name":"net.sf.ehcache.Cache",
  "allDeclaredFields":true,
  "allDeclaredMethods":true,
  "allPublicMethods":true,
  "allDeclaredConstructors":true
},
{
  "name":"net.sf.ehcache.Ehcache",
  "allDeclaredMethods":true,
  "allPublicMethods":true
},
{
  "name":"net.sf.ehcache.config.CacheConfiguration",
  "allPublicMethods":true,
  "allDeclaredConstructors":true
},
{
  "name":"net.sf.ehcache.config.Configuration",
  "allPublicMethods":true
},
{
  "name":"net.sf.ehcache.event.RegisteredEventListeners",
  "allDeclaredFields":true,
  "allPublicMethods":true
},
{
  "name":"net.sf.ehcache.statistics.extended.ExtendedStatisticsImpl$2",
  "allDeclaredMethods":true
},
{
  "name":"net.sf.ehcache.statistics.extended.ExtendedStatisticsImpl$3",
  "allDeclaredMethods":true
},
{
  "name":"net.sf.ehcache.store.AbstractStore",
  "allDeclaredFields":true
},
{
  "name":"net.sf.ehcache.store.DefaultElementValueComparator",
  "methods":[{"name":"<init>","parameterTypes":["net.sf.ehcache.config.CacheConfiguration"] }]
},
{
  "name":"net.sf.ehcache.store.MemoryStore",
  "allDeclaredFields":true,
  "allPublicMethods":true
},
{
  "name":"net.sf.ehcache.store.StoreListener",
  "allDeclaredMethods":true,
  "allPublicMethods":true
},
{
  "name":"net.sf.ehcache.store.chm.SelectableConcurrentHashMap",
  "allDeclaredFields":true,
  "allPublicMethods":true
},
{
  "name":"net.sf.ehcache.store.compound.ReadWriteSerializationCopyStrategy",
  "methods":[{"name":"<init>","parameterTypes":[] }]
},
{
  "name":"net.sf.ehcache.terracotta.InternalEhcache",
  "allDeclaredMethods":true,
  "allPublicMethods":true
},
{
  "name":"net.sf.ehcache.transaction.manager.DefaultTransactionManagerLookup",
  "methods":[{"name":"<init>","parameterTypes":[] }]
},
{
  "name":"net.sf.ehcache.writer.writethrough.WriteThroughManager",
  "allDeclaredFields":true,
  "allPublicMethods":true
}
```

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. gradlew build
2. java -cp %CP% ^
  -agentlib:native-image-agent=config-merge-dir=%GRAAL_AGENT_PATH% ^
  %MAINCLASS% > agent-output.txt
3.native-image ^
        --no-server ^
        --no-fallback ^
        --initialize-at-build-time=com.sun.jmx.remote,org.apache.xerces.util.XMLChar,org.apache.xerces.xni.NamespaceContext ^
        --initialize-at-build-time=javax.xml.datatype.DatatypeFactory ^
        --initialize-at-build-time=net.sf.ehcache.transaction ^
        -H:ConfigurationFileDirectories=C:\dev\apps\META-INF\native-image ^
        -H:Name=%ARTIFACT%-agent ^
        -H:+TraceClassInitialization ^
        -H:+RemoveSaturatedTypeFlows ^
        -H:+ReportExceptionStackTraces ^
        -Dspring.native.remove-yaml-support=true ^
        -Dspring.xml.ignore=true ^
        -Dspring.spel.ignore=true ^
        -Dspring.native.remove-jmx-support=true ^
        -Dspring.native.verify=true ^
        -cp %CP% %MAINCLASS%

**Describe GraalVM and your environment:**
 - GraalVM version : 20.3.0-dev
 - JDK major version: 1.8
 - OS: Windows 10
 - Architecture: Intel

```

[2020-08-21 14:42:29.739-0500] [APP] ERROR 28240 --- [main] net.sf.ehcache.config.Configuration      : could not instantiate transaction manager lookup class: net.sf.ehcache.transaction.manager.DefaultTransactionManagerLookup
java.lang.InstantiationException: Type `net.sf.ehcache.transaction.manager.DefaultTransactionManagerLookup` can not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.
        at java.lang.Class.newInstance(DynamicHub.java:847)
        at net.sf.ehcache.config.Configuration$RuntimeCfg.<init>(Configuration.java:1222)
        at net.sf.ehcache.config.Configuration.setupFor(Configuration.java:247)
        at net.sf.ehcache.CacheManager.doInit(CacheManager.java:439)
        at net.sf.ehcache.CacheManager.init(CacheManager.java:398)
        at net.sf.ehcache.CacheManager.<init>(CacheManager.java:374)
        at com.vue.labs.LocalAbeServiceSpringConfig.getScoreReportCache(LocalAbeServiceSpringConfig.java:144)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:650)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:483)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1338)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1177)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:557)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:226)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1304)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1224)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:884)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:788)
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:227)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1358)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1204)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:557)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:226)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1304)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1224)
        at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:884)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:788)
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:227)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1358)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1204)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:557)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:226)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:893)
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:879)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:551)
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:143)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:758)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:750)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1237)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226)
```

**Comment by oubidar-Abderrahim** [comment]
Could you provide the reproducer please

**Comment by sendev1** [comment]
@oubidar-Abderrahim I am sorry, I would not able to share the code that I am currently working on due to restrictions. Please let me know if you need any other information.

**Comment by sendev1** [comment]
I noticed this warning `WARNING: Could not register reflection metadata for net.sf.ehcache.transaction.manager.DefaultTransactionManagerLookup. Reason: java.lang.NoClassDefFoundError: javax/transaction/TransactionManager.` during image build after trying so many things and that gave me some clue. For some reason the DefaultTransactionManagerLookup was looking for TransactionManager, but it was not required when it ran in spring boot mode. So I added this dependency `implementation group: 'javax.transaction', name: 'jta', version: '1.1'` and no longer getting the `Type net.sf.ehcache.transaction.manager.DefaultTransactionManagerLookupcan not be instantiated reflectively as it does not have a no-parameter constructor or the no-parameter constructor has not been added explicitly to the native image.` error.
